programs/stake/src/stake_state.rs

```diff
@@ -842,6 +821,84 @@ impl<'a> StakeAccount for KeyedAccount<'a> {
+/// by the caller.
+struct ValidatedDelegatedInfo {
+    stake_amount: u64,
+}
+
+/// Ensure the stake delegation amount is valid.  This checks that the account meets the minimum
+/// balance requirements of delegated stake.  If not, return an error.
+fn validate_delegated_amount(
+    account: &KeyedAccount,
+    meta: &Meta,
+) -> Result<ValidatedDelegatedInfo, InstructionError> {
+    let stake_amount = account.lamports()?.saturating_sub(meta.rent_exempt_reserve); // can't stake the rent
+    Ok(ValidatedDelegatedInfo { stake_amount })
+}
+
+/// After calling `validate_split_amount()`, this struct contains calculated values that are used
+/// by the caller.
+#[derive(Copy, Clone, Debug, Default)]
+struct ValidatedSplitInfo {
+    source_remaining_balance: u64,
+    destination_rent_exempt_reserve: u64,
+}
+
+/// Ensure the split amount is valid.  This checks the source and destination accounts meet the
+/// minimum balance requirements, which is the rent exempt reserve plus the minimum stake
+/// delegation, and that the source account has enough lamports for the request split amount.  If
+/// not, return an error.
+fn validate_split_amount(
+    source_account: &KeyedAccount,
+    destination_account: &KeyedAccount,
+    lamports: u64,
+    source_meta: &Meta,
+) -> Result<ValidatedSplitInfo, InstructionError> {
+    let source_lamports = source_account.lamports()?;
+
+    // verify enough account lamports
+    if lamports > source_lamports {
+        return Err(InstructionError::InsufficientFunds);
+    }
+
+    // verify that the source account still has enough lamports left:
+    // EITHER at least the minimum balance, OR zero (in this case the source
+    // account is transferring all lamports to new destination account, and the source
+    // account will be closed)
+    let source_minimum_balance = source_meta
+        .rent_exempt_reserve
+        .saturating_add(MINIMUM_STAKE_DELEGATION);
+    let source_remaining_balance = source_lamports.saturating_sub(lamports);
+    if source_remaining_balance == 0 {
+        // full amount is a withdrawal
+        // nothing to do here
+    } else if source_remaining_balance < source_minimum_balance {
+        // the remaining balance is too low to do the split
+        return Err(InstructionError::InsufficientFunds);
+    } else {
+        // all clear!
+        // nothing to do here
+    }
+
+    // verify split amount is at least the rent exempt reserve PLUS the minimum stake delegation
+    let destination_rent_exempt_reserve = calculate_split_rent_exempt_reserve(
+        source_meta.rent_exempt_reserve,
+        source_account.data_len()? as u64,
+        destination_account.data_len()? as u64,
+    );
+    let destination_minimum_balance =
+        destination_rent_exempt_reserve.saturating_add(MINIMUM_STAKE_DELEGATION);
+    if lamports < destination_minimum_balance {
+        return Err(InstructionError::InsufficientFunds);
+    }
```

Comment:
Please correct me if I'm wrong, but I think this could incorrectly fail if the destination has some lamports already.  For example, I can `transfer` the rent exempt reserve into `split`, and then separately `split` `MINIMUM_STAKE_DELEGATION` into it, and that would fail

Reply:
Oh that's a great catch, Jon. We seem to be missing `.saturating_sub(split.lamports()?)` from the old L628 and L680 in the destination_minimum_balance calculation
Sorry for missing that @brooksprumo !

Reply:
Will fix. Phew! Glad y'all caught this. I'll also add a test for this.

Reply:
I'm also a little surprised that existing tests didn't catch this, as we used to have a bunch of cases for destination accounts with existing lamport balances. I'll take a peek.

Reply:
Yeah this really surprised me too, which is why I wasn't sure about my comment :sweat_smile:

Reply:
😞 looks like just a gap in the testing. All our tests with various destination starting-balance cases are for splitting too much or just enough. @brooksprumo , how about I pr those missing test cases (probably an extension of [this test](https://github.com/solana-labs/solana/blob/2e3eafaa116fe2f217767af4412463103c77ae3c/programs/stake/src/stake_state.rs#L4049)), and then you can rebase this PR to sanity check?

Reply:
This issue has now been resolved in this PR. Tests have been added to PR #23303. Once that PR is merged, I'll rebase this PR to confirm no behavior breakages.